docs/blocks/states.md

```diff
@@ -1,54 +1,32 @@
+A blockstate property is described by an instance of `Property<?>`. For example, slabs use an `EnumProperty<SlabType>`. Other examples include `instrument` (`EnumProperty<NoteBlockInstrument>`), `facing` (`DirectionProperty`), or `powered` (`BooleanProperty`). Each property has the value of the type `T` parametrized by `Property<T>`.
 
-```java
-switch (meta) {
-  case 0: { ... } // south and on the lower half of the block
-  case 1: { ... } // south on the upper side of the block
-  case 2: { ... } // north and on the lower half of the block
-  case 3: { ... } // north and on the upper half of the block
-  // ... etc. ...
-}
-```
-
-Because the numbers carry no meaning by themselves, no one could know what they represent unless they had access to the source code and comments.
-
-Introduction of States
----------------------------------------
-
-In Minecraft 1.8 and above, the metadata system, along with the block ID system, was deprecated and eventually replaced with the **block state system**. The block state system abstracts out the details of the block's properties from the other behaviors of the block.
-
-Each *property* of a block is described by an instance of `Property<?>`. Examples of block properties include instruments (`EnumProperty<NoteBlockInstrument>`), facing (`DirectionProperty`), poweredness (`Property<Boolean>`), etc. Each property has the value of the type `T` parametrized by `Property<T>`.
+A unique pair can be constructed from the `Block` and a map of every `Property<?>` of that block to their associated values. This unique pair is called a blockstate, represented by the `BlockState` class. As with blocks themselves, every blockstate exists exactly once in memory.
 
-A unique pair can be constructed from the `Block` and a map of the `Property<?>` to their associated values. This unique pair is called a `BlockState`.
+The `BlockState` system is a flexible and powerful system, but it also has limitations. `BlockState`s are immutable, and all combinations of their properties are generated on startup of the game. This means that having a `BlockState` with many properties and possible values will slow down the loading of the game, take up extra RAM, and may end up making your code confusing.
 
-The previous system of meaningless metadata values were replaced by a system of block properties, which are easier to interpret and deal with. Previously, a stone button which is facing east and is powered or held down is represented by "`minecraft:stone_button` with metadata `9`. Now, this is represented by "`minecraft:stone_button[facing=east,powered=true]`".
+Immutability has its upsides as well, though. For example, `==` can and should be used when comparing `BlockState`s. Another upside is that, using `BlockState#setValue(Property<?>)`, you can very easily go from one state to another. This does not actually create a new `BlockState`; it returns the previously generated state with that value instead.
```

Comment:
This is a weird pros and cons list which doesn't really help me understand the benefits of `BlockState` compared to something like `BlockEntity`.